File: Team_PageFault/minidb/txn/lock_manager.py
"""Lock manager implementing Strict Two-Phase Locking (Strict 2PL).

* Two lock modes: SHARED (read) and EXCLUSIVE (write).
* Compatibility: S/S compatible; everything else conflicts.
* Strict 2PL: locks are acquired as needed (growing phase) and *all* released
  together at commit/abort (shrinking phase happens atomically at end), which
  gives serializable isolation and recoverable schedules.
* Deadlock detection: before a transaction waits, we add its edges to a
  wait-for graph and run cycle detection. If waiting would create a cycle, the
  requesting transaction is chosen as the victim and aborted (raises
  ``DeadlockError``), breaking the cycle.

The manager is thread-safe so the concurrency demo can run real transactions in
separate threads.
"""

import logging
import threading
from collections import defaultdict

log = logging.getLogger("minidb.locks")

SHARED = "S"
EXCLUSIVE = "X"


# RuntimeError subclass so the auto-commit path's `except RuntimeError` catches it.
class DeadlockError(RuntimeError):
    pass


class _LockEntry:
    def __init__(self):
        self.mode = None                 # current granted mode (S or X)
        self.holders = set()             # txn_ids holding the lock
        self.queue = []                  # waiting (txn_id, mode) in arrival order


class LockManager:
    def __init__(self):
        self._mutex = threading.RLock()
        self._cv = threading.Condition(self._mutex)
        self._locks = defaultdict(_LockEntry)      # resource -> _LockEntry
        self._held = defaultdict(dict)             # txn_id -> {resource: mode}
        self._waits_for = defaultdict(set)         # txn_id -> set(txn_id)

    def acquire(self, txn_id, resource, mode):
        with self._cv:
            # Fast path: already hold a strong-enough lock.
            held_mode = self._held[txn_id].get(resource)
            if held_mode == EXCLUSIVE or held_mode == mode:
                return
            while not self._compatible(resource, txn_id, mode):
                holders = self._locks[resource].holders - {txn_id}
                self._waits_for[txn_id] = set(holders)
                if self._creates_cycle(txn_id):
                    self._waits_for.pop(txn_id, None)
                    log.debug("deadlock: txn %s waiting on %s would cycle; aborting it",
                              txn_id, resource)
                    raise DeadlockError(
                        f"transaction {txn_id} aborted to break a deadlock")
                self._cv.wait()
            self._waits_for.pop(txn_id, None)
            entry = self._locks[resource]
            entry.holders.add(txn_id)
            # reached only once compatible, so an X request is now the sole holder.
            if mode == EXCLUSIVE:
                entry.mode = EXCLUSIVE
            elif entry.mode is None:
                entry.mode = SHARED
            self._held[txn_id][resource] = (
                EXCLUSIVE if (mode == EXCLUSIVE or held_mode == EXCLUSIVE) else SHARED)

    def _compatible(self, resource, txn_id, mode):
        entry = self._locks[resource]
        others = entry.holders - {txn_id}
        if not others:
            return True
        if mode == SHARED and entry.mode == SHARED:
            return True            # S compatible with existing S holders
        return False               # any X involvement conflicts

    def release_all(self, txn_id):
        """Release every lock held by ``txn_id`` (commit/abort)."""
        with self._cv:
            for resource in list(self._held.get(txn_id, {})):
                entry = self._locks[resource]
                entry.holders.discard(txn_id)
                if not entry.holders:
                    entry.mode = None
            self._held.pop(txn_id, None)
            self._waits_for.pop(txn_id, None)
            for s in self._waits_for.values():
                s.discard(txn_id)
            # notify_all: every waiter must re-check, else a stale wait-for edge
            # can fake a deadlock.
            self._cv.notify_all()

    def _creates_cycle(self, start):
        """DFS over the wait-for graph to detect a cycle reachable from start."""
        visited = set()
        stack = [start]
        while stack:
            node = stack.pop()
            for nxt in self._waits_for.get(node, ()):
                if nxt == start:
                    return True
                if nxt not in visited:
                    visited.add(nxt)
                    stack.append(nxt)
        return False

File: MiniDB_Projects/Team_THE_BOYS/src/transaction/lock_manager.cpp
#include "transaction/lock_manager.h"

namespace minidb {

bool LockManager::CanGrant(const LockEntry& entry, LockMode mode, int txn_id) const {
    if (entry.holder == -1) return true;
    if (entry.holder == txn_id) return true;
    if (entry.granted == LockMode::SHARED && mode == LockMode::SHARED) return true;
    return false;
}

void LockManager::Grant(LockEntry& entry, int txn_id, LockMode mode) {
    if (entry.holder == -1) {
        entry.holder = txn_id;
        entry.granted = mode;
    } else if (entry.granted == LockMode::SHARED && mode == LockMode::EXCLUSIVE) {
        entry.granted = LockMode::EXCLUSIVE;
    }
    txn_locks_[txn_id].push_back(entry.resource);
}

bool LockManager::Lock(const std::string& resource, LockMode mode, int txn_id) {
    std::lock_guard<std::mutex> guard(mutex_);
    auto& entry = locks_[resource];
    entry.resource = resource;

    if (CanGrant(entry, mode, txn_id)) {
        Grant(entry, txn_id, mode);
        last_failure_deadlock_ = false;
        return true;
    }

    waits_on_[txn_id] = entry.holder;
    if (HasDeadlock(txn_id)) {
        waits_on_.erase(txn_id);
        last_failure_deadlock_ = true;
        return false;
    }
    last_failure_deadlock_ = false;
    return false;
}

void LockManager::UnlockAll(int txn_id) {
    std::lock_guard<std::mutex> guard(mutex_);
    auto it = txn_locks_.find(txn_id);
    if (it == txn_locks_.end()) return;
    for (const auto& resource : it->second) {
        locks_.erase(resource);
    }
    txn_locks_.erase(txn_id);
    waits_on_.erase(txn_id);
}

bool LockManager::HasDeadlock(int txn_id) const {
    int current = txn_id;
    for (int i = 0; i < 32; ++i) {
        auto it = waits_on_.find(current);
        if (it == waits_on_.end()) return false;
        if (it->second == txn_id) return true;
        current = it->second;
    }
    return false;
}

std::string LockManager::DeadlockVictim() const {
    if (waits_on_.empty()) return "";
    return std::to_string(waits_on_.begin()->first);
}

}  // namespace minidb
